netmaster/resources/epgresource.go

```diff
@@ -0,0 +1,298 @@
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/contiv/netplugin/core"
+	"github.com/contiv/netplugin/netmaster/mastercfg"
+	"github.com/jainvipin/bitset"
+)
+
+const (
+	// AutoEPGResource is the name of the resource, for storing state.
+	AutoEPGResource = "auto-epg"
+)
+
+const (
+	epgResourceConfigPathPrefix = mastercfg.StateConfigPath + AutoEPGResource + "/"
+	epgResourceConfigPath       = epgResourceConfigPathPrefix + "%s"
+	epgResourceOperPathPrefix   = mastercfg.StateOperPath + AutoEPGResource + "/"
+	epgResourceOperPath         = epgResourceOperPathPrefix + "%s"
+)
+
+// AutoEPGCfgResource implements the Resource interface for an 'auto-epg' resource.
+// 'auto-epg' resource allocates an EPG ID from a range specified
+// at time of resource instantiation
+type AutoEPGCfgResource struct {
+	core.CommonState
+	EPGs *bitset.BitSet `json:"epgs"`
+}
+
+// Write the state.
+func (r *AutoEPGCfgResource) Write() error {
+	key := fmt.Sprintf(epgResourceConfigPath, r.ID)
+	return r.StateDriver.WriteState(key, r, json.Marshal)
+}
+
+// Read the state.
+func (r *AutoEPGCfgResource) Read(id string) error {
+	key := fmt.Sprintf(epgResourceConfigPath, id)
+	return r.StateDriver.ReadState(key, r, json.Unmarshal)
+}
+
+// Clear the state.
+func (r *AutoEPGCfgResource) Clear() error {
+	key := fmt.Sprintf(epgResourceConfigPath, r.ID)
+	return r.StateDriver.ClearState(key)
+}
+
+// ReadAll the state for this resource.
+func (r *AutoEPGCfgResource) ReadAll() ([]core.State, error) {
+	return r.StateDriver.ReadAllState(epgResourceConfigPathPrefix, r,
+		json.Unmarshal)
+}
+
+// Init the Resource. Requires a *bitset.BitSet.
+func (r *AutoEPGCfgResource) Init(rsrcCfg interface{}) error {
+	var ok bool
+	r.EPGs, ok = rsrcCfg.(*bitset.BitSet)
+	if !ok {
+		return core.Errorf("Invalid type for EPG resource config")
+	}
+	err := r.Write()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			r.Clear()
+		}
+	}()
+
+	oper := &AutoEPGOperResource{FreeEPGs: r.EPGs}
+	oper.StateDriver = r.StateDriver
+	oper.ID = r.ID
+	err = oper.Write()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Deinit the resource.
+func (r *AutoEPGCfgResource) Deinit() {
+	oper := &AutoEPGOperResource{}
+	oper.StateDriver = r.StateDriver
+	err := oper.Read(r.ID)
+	if err != nil {
+		// continue cleanup
+	} else {
+		err = oper.Clear()
+		if err != nil {
+			// continue cleanup
+		}
+	}
+
+	r.Clear()
+}
+
+// Reinit the Resource. Requires a *bitset.BitSet.
+func (r *AutoEPGCfgResource) Reinit(rsrcCfg interface{}) error {
```

Comment:
again implementation may not be needed if we are not letting the range configurable